File: 0018-4sum/0018-4sum.cpp
class Solution
{
public:
    vector<vector<int>> fourSum(vector<int> &nums, int target)
    {
        int i = 0;
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        // if(nums.size()<4)return ans;
        int n=nums.size();
       for(int i=0;i<n-2;i++)
        {
           //cout<<i<<endl;
            if(i>0 and nums[i]==nums[i-1])
            continue;
            
           // int j = i + 1;
            for(int j=i+1;j<n-1;j++)
            {
                if (j > i + 1 and nums[j] == nums[j - 1])
                    continue;
                long long int v1 = (long long int)nums[i];
                long long int v2 = (long long int)nums[j];
                long long int targetsum = (long long int)target - v1 - v2;
                long long int k = (long long int)j + 1;
                long long int l = (long long int)nums.size() - 1;

                while (k < l)
                {
                    vector<int> temp;
                    long long int v3 = (long long int)nums[k];
                    long long int v4 = (long long int)nums[l];
                    // cout<<"["<<v1<<","<<v2<<","<<v3<<","<<v4<<"]"<<endl;
                    if (v3 + v4 == targetsum)
                    {
                        temp.push_back(v1);
                        temp.push_back(v2);
                        temp.push_back(nums[k]);
                        temp.push_back(nums[l]);
                        ans.push_back(temp);
                        long long int kv = v3;
                        long long int lv = v4;
                        while (k < l && kv == nums[k])
                            k++;
                        while (l > k && lv == nums[l])
                            l--;
                    }
                    else if (v3 + v4 < targetsum)
                        k++;
                    else
                        l--;
                }
                //j++;
            }
            //i++;
        }
        return ans;
    }
};

File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>>ans=new ArrayList<>();
        int n=nums.length;
        for(int i=0;i<nums.length-3;i++){
            if(i>0 && nums[i-1]==nums[i])continue;
            for(int j=i+1;j<nums.length-2;j++){
                if(j>i+1 && nums[j-1]==nums[j])continue;
                long sum=(long)target-nums[i]-nums[j];
                int l=j+1;
                int k=n-1;
                while(l<k){
                    if((long) nums[l]+nums[k]==sum){
                        List<Integer>ls=new ArrayList<>();
                        ls.add(nums[i]);
                        ls.add(nums[j]);
                        ls.add(nums[k]);
                        ls.add(nums[l]);
                        ans.add(ls);
                        l++;
                        k--;
                        while(l<k && nums[l]==nums[l-1])l++;
                        while(l<k && nums[k]==nums[k+1])k--;
                    }
                    else if((long) nums[l]+nums[k]<sum){
                        l++;
                    }
                    else{
                        k--;
                    }
                }
            }
          
        }
        return ans;
    }
}